File: ui/desktop/src/utils/envCompat.ts
/**
 * Environment Variable Compatibility Layer for AGIME/Goose Migration
 *
 * This module provides dual-prefix support for configuration during
 * the brand migration from "Goose" to "AGIME". Configuration can be accessed
 * using either AGIME_* prefix (preferred) or GOOSE_* prefix (legacy).
 *
 * ## Priority Order
 * 1. AGIME_ prefix (new, preferred)
 * 2. GOOSE_ prefix (legacy, fallback)
 *
 * ## Usage Examples
 * ```typescript
 * import { getConfigCompat, getEnvCompat, configCompatExists } from '@/utils/envCompat';
 *
 * // Get from window.appConfig with dual-prefix support
 * const provider = getConfigCompat('PROVIDER'); // Checks AGIME_PROVIDER, then GOOSE_PROVIDER
 *
 * // Get from process.env with dual-prefix support
 * const apiKey = getEnvCompat('API_KEY'); // Checks AGIME_API_KEY, then GOOSE_API_KEY
 *
 * // Get with typed default
 * const mode = getConfigCompatTyped<string>('MODE', 'auto');
 * ```
 */

/** The new AGIME prefix for configuration (preferred) */
export const AGIME_PREFIX = 'AGIME_';

/** The legacy GOOSE prefix for configuration (fallback) */
export const GOOSE_PREFIX = 'GOOSE_';

/**
 * Get a configuration value from window.appConfig with dual-prefix support.
 *
 * Tries AGIME_ prefix first, then falls back to GOOSE_ prefix.
 *
 * @param key - The key without prefix (e.g., "PROVIDER" not "GOOSE_PROVIDER")
 * @returns The configuration value, or undefined if not found
 *
 * @example
 * // With GOOSE_MODE=auto in config:
 * getConfigCompat('MODE') // returns 'auto'
 *
 * // With both AGIME_MODE=manual and GOOSE_MODE=auto:
 * getConfigCompat('MODE') // returns 'manual' (AGIME takes priority)
 */
export function getConfigCompat(key: string): unknown {
  // Check if window.appConfig is available (Electron environment)
  if (typeof window !== 'undefined' && window.appConfig) {
    // Tier 1: Try AGIME_ prefix (preferred)
    const agimeKey = `${AGIME_PREFIX}${key}`;
    const agimeValue = window.appConfig.get?.(agimeKey);
    if (agimeValue !== undefined && agimeValue !== null) {
      return agimeValue;
    }

    // Tier 2: Try GOOSE_ prefix (legacy fallback)
    const gooseKey = `${GOOSE_PREFIX}${key}`;
    const gooseValue = window.appConfig.get?.(gooseKey);
    if (gooseValue !== undefined && gooseValue !== null) {
      return gooseValue;
    }
  }

  return undefined;
}

/**
 * Get a configuration value with type safety and optional default.
 *
 * @param key - The key without prefix
 * @param defaultValue - Optional default value if not found
 * @returns The typed configuration value, or default/undefined if not found
 */
export function getConfigCompatTyped<T>(key: string, defaultValue?: T): T | undefined {
  const value = getConfigCompat(key);
  if (value !== undefined) {
    return value as T;
  }
  return defaultValue;
}

/**
 * Get an environment variable from process.env with dual-prefix support.
 *
 * @param key - The key without prefix
 * @returns The environment variable value, or undefined if not found
 */
export function getEnvCompat(key: string): string | undefined {
  // Check if process.env is available
  if (typeof process !== 'undefined' && process.env) {
    // Tier 1: Try AGIME_ prefix (preferred)
    const agimeKey = `${AGIME_PREFIX}${key}`;
    const agimeValue = process.env[agimeKey];
    if (agimeValue !== undefined && agimeValue !== '') {
      return agimeValue;
    }

    // Tier 2: Try GOOSE_ prefix (legacy fallback)
    const gooseKey = `${GOOSE_PREFIX}${key}`;
    const gooseValue = process.env[gooseKey];
    if (gooseValue !== undefined && gooseValue !== '') {
      return gooseValue;
    }
  }

  return undefined;
}

/**
 * Get an environment variable with a default value.
 *
 * @param key - The key without prefix
 * @param defaultValue - Default value if not found
 * @returns The environment variable value, or the default
 */
export function getEnvCompatOr(key: string, defaultValue: string): string {
  return getEnvCompat(key) ?? defaultValue;
}

/**
 * Get a configuration value from either appConfig or process.env with dual-prefix support.
 *
 * Priority:
 * 1. window.appConfig with AGIME_ prefix
 * 2. window.appConfig with GOOSE_ prefix
 * 3. process.env with AGIME_ prefix
 * 4. process.env with GOOSE_ prefix
 *
 * @param key - The key without prefix
 * @returns The configuration value as string, or undefined if not found
 */
export function getAnyConfigCompat(key: string): string | undefined {
  // First try appConfig
  const appConfigValue = getConfigCompat(key);
  if (appConfigValue !== undefined) {
    return String(appConfigValue);
  }

  // Then try process.env
  return getEnvCompat(key);
}

/**
 * Check if a configuration exists under either prefix.
 *
 * @param key - The key without prefix
 * @returns true if the configuration exists under AGIME_ or GOOSE_ prefix
 */
export function configCompatExists(key: string): boolean {
  if (typeof window !== 'undefined' && window.appConfig) {
    const agimeKey = `${AGIME_PREFIX}${key}`;
    const gooseKey = `${GOOSE_PREFIX}${key}`;

    const agimeValue = window.appConfig.get?.(agimeKey);
    const gooseValue = window.appConfig.get?.(gooseKey);

    if (agimeValue !== undefined && agimeValue !== null) return true;
    if (gooseValue !== undefined && gooseValue !== null) return true;
  }

  return false;
}

/**
 * Check if an environment variable exists under either prefix.
 *
 * @param key - The key without prefix
 * @returns true if the env var exists under AGIME_ or GOOSE_ prefix
 */
export function envCompatExists(key: string): boolean {
  if (typeof process !== 'undefined' && process.env) {
    const agimeKey = `${AGIME_PREFIX}${key}`;
    const gooseKey = `${GOOSE_PREFIX}${key}`;

    if (process.env[agimeKey] !== undefined && process.env[agimeKey] !== '') return true;
    if (process.env[gooseKey] !== undefined && process.env[gooseKey] !== '') return true;
  }

  return false;
}

/**
 * Get the actual key name that was found (for debugging/logging).
 *
 * @param key - The key without prefix
 * @returns Object with the actual prefixed key name, value, and source
 */
export function getConfigCompatWithSource(
  key: string
): { fullKey: string; value: unknown; source: 'appConfig' | 'env' } | undefined {
  // Check appConfig first
  if (typeof window !== 'undefined' && window.appConfig) {
    const agimeKey = `${AGIME_PREFIX}${key}`;
    const agimeValue = window.appConfig.get?.(agimeKey);
    if (agimeValue !== undefined && agimeValue !== null) {
      return { fullKey: agimeKey, value: agimeValue, source: 'appConfig' };
    }

    const gooseKey = `${GOOSE_PREFIX}${key}`;
    const gooseValue = window.appConfig.get?.(gooseKey);
    if (gooseValue !== undefined && gooseValue !== null) {
      return { fullKey: gooseKey, value: gooseValue, source: 'appConfig' };
    }
  }

  // Check process.env
  if (typeof process !== 'undefined' && process.env) {
    const agimeKey = `${AGIME_PREFIX}${key}`;
    const agimeValue = process.env[agimeKey];
    if (agimeValue !== undefined && agimeValue !== '') {
      return { fullKey: agimeKey, value: agimeValue, source: 'env' };
    }

    const gooseKey = `${GOOSE_PREFIX}${key}`;
    const gooseValue = process.env[gooseKey];
    if (gooseValue !== undefined && gooseValue !== '') {
      return { fullKey: gooseKey, value: gooseValue, source: 'env' };
    }
  }

  return undefined;
}

/**
 * Convert a legacy GOOSE_ prefixed key to the new AGIME_ prefix.
 *
 * @param key - A key that may have GOOSE_ prefix
 * @returns The key with AGIME_ prefix if it had GOOSE_ prefix, otherwise unchanged
 */
export function migrateKeyToAgime(key: string): string {
  if (key.startsWith(GOOSE_PREFIX)) {
    return `${AGIME_PREFIX}${key.slice(GOOSE_PREFIX.length)}`;
  }
  return key;
}

/**
 * Check if a full key (with prefix) is using the legacy GOOSE_ prefix.
 */
export function isLegacyKey(key: string): boolean {
  return key.startsWith(GOOSE_PREFIX);
}

/**
 * Check if a full key (with prefix) is using the new AGIME_ prefix.
 */
export function isAgimeKey(key: string): boolean {
  return key.startsWith(AGIME_PREFIX);
}

/**
 * Get the base key without any prefix.
 *
 * @param key - A key that may have AGIME_ or GOOSE_ prefix
 * @returns The key without prefix
 */
export function stripPrefix(key: string): string {
  if (key.startsWith(AGIME_PREFIX)) {
    return key.slice(AGIME_PREFIX.length);
  }
  if (key.startsWith(GOOSE_PREFIX)) {
    return key.slice(GOOSE_PREFIX.length);
  }
  return key;
}

/**
 * Build a full key with the preferred AGIME_ prefix.
 *
 * @param baseKey - The key without prefix
 * @returns The key with AGIME_ prefix
 */
export function buildAgimeKey(baseKey: string): string {
  const stripped = stripPrefix(baseKey);
  return `${AGIME_PREFIX}${stripped}`;
}

/**
 * Build a full key with the legacy GOOSE_ prefix.
 *
 * @param baseKey - The key without prefix
 * @returns The key with GOOSE_ prefix
 */
export function buildGooseKey(baseKey: string): string {
  const stripped = stripPrefix(baseKey);
  return `${GOOSE_PREFIX}${stripped}`;
}

// Re-export for backward compatibility with existing code that directly uses GOOSE_ keys
// These constants can be used to maintain compatibility while transitioning
export const LEGACY_CONFIG_KEYS = {
  PROVIDER: 'GOOSE_PROVIDER',
  MODEL: 'GOOSE_MODEL',
  MODE: 'GOOSE_MODE',
  WORKING_DIR: 'GOOSE_WORKING_DIR',
  API_HOST: 'GOOSE_API_HOST',
  DEFAULT_PROVIDER: 'GOOSE_DEFAULT_PROVIDER',
  DEFAULT_MODEL: 'GOOSE_DEFAULT_MODEL',
  PREDEFINED_MODELS: 'GOOSE_PREDEFINED_MODELS',
  BASE_URL_SHARE: 'GOOSE_BASE_URL_SHARE',
  VERSION: 'GOOSE_VERSION',
  EXTERNAL_BACKEND: 'GOOSE_EXTERNAL_BACKEND',
  ALLOWLIST_WARNING: 'GOOSE_ALLOWLIST_WARNING',
  PATH_ROOT: 'GOOSE_PATH_ROOT',
  ALLOWLIST: 'GOOSE_ALLOWLIST',
  MAX_TURNS: 'GOOSE_MAX_TURNS',
  ENABLE_ROUTER: 'GOOSE_ENABLE_ROUTER',
  LEAD_MODEL: 'GOOSE_LEAD_MODEL',
  LEAD_PROVIDER: 'GOOSE_LEAD_PROVIDER',
  LEAD_TURNS: 'GOOSE_LEAD_TURNS',
  LEAD_FAILURE_THRESHOLD: 'GOOSE_LEAD_FAILURE_THRESHOLD',
  LEAD_FALLBACK_TURNS: 'GOOSE_LEAD_FALLBACK_TURNS',
  SERVER_MEMORY: 'GOOSE_SERVER__MEMORY',
  SERVER_COMPUTER_CONTROLLER: 'GOOSE_SERVER__COMPUTER_CONTROLLER',
} as const;

// New AGIME_ equivalents
export const AGIME_CONFIG_KEYS = {
  PROVIDER: 'AGIME_PROVIDER',
  MODEL: 'AGIME_MODEL',
  MODE: 'AGIME_MODE',
  WORKING_DIR: 'AGIME_WORKING_DIR',
  API_HOST: 'AGIME_API_HOST',
  DEFAULT_PROVIDER: 'AGIME_DEFAULT_PROVIDER',
  DEFAULT_MODEL: 'AGIME_DEFAULT_MODEL',
  PREDEFINED_MODELS: 'AGIME_PREDEFINED_MODELS',
  BASE_URL_SHARE: 'AGIME_BASE_URL_SHARE',
  VERSION: 'AGIME_VERSION',
  EXTERNAL_BACKEND: 'AGIME_EXTERNAL_BACKEND',
  ALLOWLIST_WARNING: 'AGIME_ALLOWLIST_WARNING',
  PATH_ROOT: 'AGIME_PATH_ROOT',
  ALLOWLIST: 'AGIME_ALLOWLIST',
  MAX_TURNS: 'AGIME_MAX_TURNS',
  ENABLE_ROUTER: 'AGIME_ENABLE_ROUTER',
  LEAD_MODEL: 'AGIME_LEAD_MODEL',
  LEAD_PROVIDER: 'AGIME_LEAD_PROVIDER',
  LEAD_TURNS: 'AGIME_LEAD_TURNS',
  LEAD_FAILURE_THRESHOLD: 'AGIME_LEAD_FAILURE_THRESHOLD',
  LEAD_FALLBACK_TURNS: 'AGIME_LEAD_FALLBACK_TURNS',
  SERVER_MEMORY: 'AGIME_SERVER__MEMORY',
  SERVER_COMPUTER_CONTROLLER: 'AGIME_SERVER__COMPUTER_CONTROLLER',
} as const;

File: crates/agime/src/config/env_compat.rs
//! Environment Variable Compatibility Layer for AGIME/Goose Migration
//!
//! This module provides dual-prefix support for environment variables during
//! the brand migration from "Goose" to "AGIME". Variables can be accessed using
//! either AGIME_* prefix (preferred) or GOOSE_* prefix (legacy).
//!
//! # Priority Order
//! 1. AGIME_ prefix (new, preferred)
//! 2. GOOSE_ prefix (legacy, fallback)
//!
//! # Usage Examples
//! ```rust,ignore
//! use agime::config::env_compat::{get_env_compat, get_env_compat_or, env_compat_exists};
//!
//! // Get an optional environment variable
//! if let Some(value) = get_env_compat("PROVIDER") {
//!     // Checks AGIME_PROVIDER first, then GOOSE_PROVIDER
//! }
//!
//! // Get with default value
//! let mode = get_env_compat_or("MODE", "auto");
//!
//! // Check if variable exists under either prefix
//! if env_compat_exists("API_KEY") {
//!     // AGIME_API_KEY or GOOSE_API_KEY is set
//! }
//! ```

use std::env;
use std::str::FromStr;

/// The new AGIME prefix for environment variables (preferred)
pub const AGIME_PREFIX: &str = "AGIME_";

/// The legacy GOOSE prefix for environment variables (fallback)
pub const GOOSE_PREFIX: &str = "GOOSE_";

/// Get an environment variable with dual-prefix support.
///
/// Tries AGIME_ prefix first, then falls back to GOOSE_ prefix.
///
/// # Arguments
/// * `key` - The key without prefix (e.g., "PROVIDER" not "GOOSE_PROVIDER")
///
/// # Returns
/// * `Some(String)` if the variable is found under either prefix
/// * `None` if not found under any prefix
///
/// # Examples
/// ```rust,ignore
/// // With GOOSE_MODE=auto set:
/// assert_eq!(get_env_compat("MODE"), Some("auto".to_string()));
///
/// // With AGIME_MODE=manual set (takes priority):
/// assert_eq!(get_env_compat("MODE"), Some("manual".to_string()));
/// ```
pub fn get_env_compat(key: &str) -> Option<String> {
    // Tier 1: Try AGIME_ prefix (preferred)
    let agime_key = format!("{}{}", AGIME_PREFIX, key);
    if let Ok(value) = env::var(&agime_key) {
        return Some(value);
    }

    // Tier 2: Try GOOSE_ prefix (legacy fallback)
    let goose_key = format!("{}{}", GOOSE_PREFIX, key);
    if let Ok(value) = env::var(&goose_key) {
        return Some(value);
    }

    None
}

/// Get an environment variable with dual-prefix support, or return a default value.
///
/// # Arguments
/// * `key` - The key without prefix
/// * `default` - Default value if not found under either prefix
///
/// # Returns
/// The environment variable value, or the default if not found
pub fn get_env_compat_or(key: &str, default: &str) -> String {
    get_env_compat(key).unwrap_or_else(|| default.to_string())
}

/// Get and parse an environment variable with dual-prefix support.
///
/// # Arguments
/// * `key` - The key without prefix
///
/// # Returns
/// * `Ok(T)` if the variable is found and successfully parsed
/// * `Err` if not found or parsing fails
///
/// # Examples
/// ```rust,ignore
/// // With GOOSE_PORT=8080:
/// let port: u16 = get_env_compat_parsed("PORT")?;
/// assert_eq!(port, 8080);
/// ```
pub fn get_env_compat_parsed<T>(key: &str) -> Result<T, Box<dyn std::error::Error + Send + Sync>>
where
    T: FromStr,
    T::Err: std::error::Error + Send + Sync + 'static,
{
    let value = get_env_compat(key).ok_or_else(|| {
        format!(
            "Environment variable not found: {}_{} or {}_{}",
            AGIME_PREFIX, key, GOOSE_PREFIX, key
        )
    })?;
    value
        .parse::<T>()
        .map_err(|e| Box::new(e) as Box<dyn std::error::Error + Send + Sync>)
}

/// Get and parse an environment variable with dual-prefix support, or return a default.
///
/// # Arguments
/// * `key` - The key without prefix
/// * `default` - Default value if not found or parsing fails
pub fn get_env_compat_parsed_or<T>(key: &str, default: T) -> T
where
    T: FromStr,
    T::Err: std::error::Error + Send + Sync + 'static,
{
    get_env_compat_parsed(key).unwrap_or(default)
}

/// Check if an environment variable exists under either prefix.
///
/// # Arguments
/// * `key` - The key without prefix
///
/// # Returns
/// `true` if the variable exists under AGIME_ or GOOSE_ prefix
pub fn env_compat_exists(key: &str) -> bool {
    let agime_key = format!("{}{}", AGIME_PREFIX, key);
    let goose_key = format!("{}{}", GOOSE_PREFIX, key);
    env::var(&agime_key).is_ok() || env::var(&goose_key).is_ok()
}

/// Get the actual key name that was found (for debugging/logging).
///
/// # Arguments
/// * `key` - The key without prefix
///
/// # Returns
/// * `Some((full_key, value))` with the actual prefixed key name and value
/// * `None` if not found
pub fn get_env_compat_with_source(key: &str) -> Option<(String, String)> {
    let agime_key = format!("{}{}", AGIME_PREFIX, key);
    if let Ok(value) = env::var(&agime_key) {
        return Some((agime_key, value));
    }

    let goose_key = format!("{}{}", GOOSE_PREFIX, key);
    if let Ok(value) = env::var(&goose_key) {
        return Some((goose_key, value));
    }

    None
}

/// Convert a legacy GOOSE_ prefixed key to the new AGIME_ prefix.
///
/// # Arguments
/// * `key` - A key that may have GOOSE_ prefix
///
/// # Returns
/// The key with AGIME_ prefix if it had GOOSE_ prefix, otherwise unchanged
pub fn migrate_key_to_agime(key: &str) -> String {
    if key.starts_with(GOOSE_PREFIX) {
        format!("{}{}", AGIME_PREFIX, &key[GOOSE_PREFIX.len()..])
    } else {
        key.to_string()
    }
}

/// Check if a full key (with prefix) is using the legacy GOOSE_ prefix.
pub fn is_legacy_key(key: &str) -> bool {
    key.starts_with(GOOSE_PREFIX)
}

/// Check if a full key (with prefix) is using the new AGIME_ prefix.
pub fn is_agime_key(key: &str) -> bool {
    key.starts_with(AGIME_PREFIX)
}

/// Macro for convenient environment variable access with dual-prefix support.
///
/// # Usage
/// ```rust,ignore
/// // Get optional value
/// let value = env_compat!("PROVIDER");
///
/// // Get with default
/// let mode = env_compat!("MODE", "auto");
/// ```
#[macro_export]
macro_rules! env_compat {
    ($key:expr) => {
        $crate::config::env_compat::get_env_compat($key)
    };
    ($key:expr, $default:expr) => {
        $crate::config::env_compat::get_env_compat_or($key, $default)
    };
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::env;

    // Note: These tests manipulate environment variables, so they should be run
    // with --test-threads=1 to avoid race conditions

    #[test]
    fn test_goose_prefix_fallback() {
        let test_key = "TEST_COMPAT_GOOSE_ONLY";
        let goose_key = format!("{}{}", GOOSE_PREFIX, test_key);

        // Clean up
        env::remove_var(&goose_key);
        env::remove_var(&format!("{}{}", AGIME_PREFIX, test_key));

        // Set only GOOSE_ version
        env::set_var(&goose_key, "goose_value");

        assert_eq!(get_env_compat(test_key), Some("goose_value".to_string()));

        // Clean up
        env::remove_var(&goose_key);
    }

    #[test]
    fn test_agime_prefix_priority() {
        let test_key = "TEST_COMPAT_BOTH";
        let agime_key = format!("{}{}", AGIME_PREFIX, test_key);
        let goose_key = format!("{}{}", GOOSE_PREFIX, test_key);

        // Clean up
        env::remove_var(&agime_key);
        env::remove_var(&goose_key);

        // Set both
        env::set_var(&goose_key, "goose_value");
        env::set_var(&agime_key, "agime_value");

        // AGIME should take priority
        assert_eq!(get_env_compat(test_key), Some("agime_value".to_string()));

        // Clean up
        env::remove_var(&agime_key);
        env::remove_var(&goose_key);
    }

    #[test]
    fn test_not_found() {
        let test_key = "TEST_COMPAT_NOT_EXISTS_12345";
        assert_eq!(get_env_compat(test_key), None);
    }

    #[test]
    fn test_default_value() {
        let test_key = "TEST_COMPAT_DEFAULT_12345";
        assert_eq!(get_env_compat_or(test_key, "default"), "default");
    }

    #[test]
    fn test_env_compat_exists() {
        let test_key = "TEST_COMPAT_EXISTS";
        let goose_key = format!("{}{}", GOOSE_PREFIX, test_key);

        env::remove_var(&goose_key);
        assert!(!env_compat_exists(test_key));

        env::set_var(&goose_key, "value");
        assert!(env_compat_exists(test_key));

        env::remove_var(&goose_key);
    }

    #[test]
    fn test_parsed_value() {
        let test_key = "TEST_COMPAT_PARSED";
        let goose_key = format!("{}{}", GOOSE_PREFIX, test_key);

        env::set_var(&goose_key, "42");
        let value: u32 = get_env_compat_parsed(test_key).unwrap();
        assert_eq!(value, 42);

        env::remove_var(&goose_key);
    }

    #[test]
    fn test_migrate_key() {
        assert_eq!(migrate_key_to_agime("GOOSE_PROVIDER"), "AGIME_PROVIDER");
        assert_eq!(migrate_key_to_agime("AGIME_PROVIDER"), "AGIME_PROVIDER");
        assert_eq!(migrate_key_to_agime("OTHER_KEY"), "OTHER_KEY");
    }

    #[test]
    fn test_is_legacy_key() {
        assert!(is_legacy_key("GOOSE_PROVIDER"));
        assert!(!is_legacy_key("AGIME_PROVIDER"));
        assert!(!is_legacy_key("OTHER_KEY"));
    }
}
